src/indexed_crate.rs

```diff
@@ -231,154 +231,139 @@ pub struct IndexedCrate<'a> {
-                structs: Vec::with_capacity(structs_len),
-                enums: Vec::with_capacity(enums_len),
-                unions: Vec::with_capacity(unions_len),
-                traits: Vec::with_capacity(traits_len),
-                modules: Vec::with_capacity(modules_len),
-                statics: Vec::with_capacity(statics_len),
-                free_consts: Vec::with_capacity(consts_len),
-                decl_macros: Vec::with_capacity(decl_macros_len),
-                proc_macros: Vec::with_capacity(proc_macros_len),
-            };
+        })
+    }
 
-            for c in collected {
-                value.free_functions.extend(c.free_functions);
-                value.structs.extend(c.structs);
-                value.enums.extend(c.enums);
-                value.unions.extend(c.unions);
-                value.traits.extend(c.traits);
-                value.modules.extend(c.modules);
-                value.statics.extend(c.statics);
-                value.free_consts.extend(c.free_consts);
-                value.decl_macros.extend(c.decl_macros);
-                value.proc_macros.extend(c.proc_macros);
+    /// Returns true if the index corresponding to the type given by
+    /// `destination_type` contains an element with the id `item_id`.
+    /// Conservatively, returns true if the destination is None or of a type
+    /// that isn't contained in the indexes.
+    pub fn contains(&self, destination_type: Option<Arc<str>>, item_id: Id) -> bool {
+        if let Some(destination_type) = destination_type {
+            match destination_type.as_ref() {
+                "Function" => self.free_functions.contains_key(&item_id),
+                "Struct" => self.structs.contains_key(&item_id),
+                "Enum" => self.enums.contains_key(&item_id),
+                "Union" => self.unions.contains_key(&item_id),
+                "Trait" => self.traits.contains_key(&item_id),
+                "ImplOwner" => {
+                    self.structs.contains_key(&item_id)
+                        || self.enums.contains_key(&item_id)
+                        || self.unions.contains_key(&item_id)
+                }
+                "Constant" => self.free_consts.contains_key(&item_id),
+                "Static" => self.statics.contains_key(&item_id),
+                "GlobalValue" => {
+                    // const or static
+                    self.free_consts.contains_key(&item_id) || self.statics.contains_key(&item_id)
+                }
+                "Macro" => self.decl_macros.contains_key(&item_id),
+                "ProcMacro"
+                | "FunctionLikeProcMacro"
+                | "AttributeProcMacro"
+                | "DeriveProcMacro" => self.proc_macros.contains_key(&item_id),
+                "Module" => self.modules.contains_key(&item_id),
+                _ => true,
             }
-
-            value
-        };
-
-        folded
+        } else {
+            true
+        }
```

Comment:
nit: The "if `None` return `true`" side seems not particularly worth it here. Consider refactoring this to push the `None` check into the caller, and have `contains()` only take a `&str`.